README.md
# RP2040-BACnetServerExample

Server example using the Raspberry PI RP2040 wifi chip

In this project we are using the [CAS BACnet stack](https://www.bacnetstack.com/) to generate a simple BACnet server with an Analog Input, Multi-state-value (MSV), and Network Port object. The analog input displays the CPU temperature and can be configured with Celsius or Fahrenheit units. The MSV shows the current mode of the built-in LED.

A BACnet client (such as the [CAS BACnet Explorer](https://store.chipkin.com/products/tools/cas-bacnet-explorer)) can be used to write to this MSV to change the mode.

The MSV allows for three possible values to be written.

- 1 = **Off**
- 2 = **On**
- 3 = **Blink** - Blinks the LED on and off at a rate of 1000 ms.
- 4 = **Fast Blink** - Blinks the LED on and off at a rate of 250 ms.

The Analog Input allows for three possible values to be written to the units.

- 1 = **Celsius** - 62
- 2 = **Fahrenheit** - 64

## Quick start

The [CAS BACnet Stack submodule](https://github.com/chipkin/BACnetServerExampleCPP/issues/8) is required for compilation.

For the example server to run properly, please enable the needed object types and features of the CAS BACNet Stack. For more details, please reference `Enabling optional functionality` and `Compiling example projects` sections of the *Quick Start Guide* (please contact Chipkin for this document).

To connect the pico to wifi, you need to update the WIFI_SSID and WIFI_PASSWORD fields in the BACnetServerExamplePico.cpp file.

## Build

1. Download and install the pico standard library and compilation tools
2. Update repository submodules
3. Create "build" folder in top of directory
4. Run "cmake .."
5. Run "make"
6. Holding the BOOTSEL button, plug the Raspberry Pi Pico W into the computer
7. Drag the compiled "BACnetServerExamplePico.uf2" file from the build directory onto the Raspberry Pi Pico W build folder

## Supported BIBBs

The CAS BACnet stack supports many other BIBBs, this minumial example only supports the following:

- **DS-RP-B**: Data Sharing - ReadProperty-B
- **DS-RPM-B**: Data Sharing - ReadPropertyMultiple-B
- **DS-WP-B**: Data Sharing - WriteProperty-B
- **DM-DDB-B**: Device and Network Management - Dynamic Device Binding-B (Who-Is)
- **DM-DOB-B**: Device and Network Management - Dynamic Object Binding-B (Who-Has)

## BACnet objects supported

The CAS BACnet stack supports many other object types this minimal example only supports the following:

- Analog Input (0)
- Device (8)
- Multi-state Value (19)
- Network Port (56)

## Device Tree

Below is the device tree.

![Preview of device tree](/device_tree.png?raw=true "Preview of device tree")

## Example of setting LED mode

![Preview of setting the MSV LED mode](/LED_demo.gif?raw=true "Preview of setting the MSV LED mode")

## Example output

The example output from the serial terminal of the ESP32

```txt
FYI: Attempting to connect to wifi SSID
FYI: Connected!
FYI: Binded to UDP socket 47808
FYI: CAS BACnet Stack version: 4.1.19.0
FYI: BACnet device created: Device instance=389007
FYI: Enabled WriteProperty for Device 389007
FYI: Enabled Read Property Multiple for Device 14
FYI: Enabling SubscribeCOV... OK
FYI: Adding AnalogInput. analogInput.instance=[2]... OK
FYI: Adding NetworkPort. networkPort.instance=[0]... OK
FYI: NetworkPort.IPAddress: 192.168.1.78
FYI: NetworkPort.IPSubnetMask: 255.255.255.0
FYI: NetworkPort.BroadcastIPAddress: 192.168.1.255
FYI: CallbackGetPropertyUInt deviceInstance=389007, objectType=8, objectInstance=389007, propertyIdentifier=62
FYI: CallbackGetPropertyUInt deviceInstance=389007, objectType=8, objectInstance=389007, propertyIdentifier=120
CallbackGetPropertyEnum deviceInstance=389007, , objectType=8, objectInstance=389007, propertyIdentifier=107 useArrayIndex=0, propertyArrayIndex=0
FYI: Sent message with 25 bytes to 192.168.1.255:47808
FYI: Sent broadcast IAm message
FYI: Entering main loop
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

```

## Tested hardware

- [Raspberry Pi Pico W](https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html)

## FAQ

### Can't build the project. *CASBACnetStackAdapter.h: No such file or directory*

Please see this issue [CASBACnetStackAdapter.h: No such file or directory](https://github.com/chipkin/ESP32-BACnetServerExample/issues/1)

### Binary size too large for FLASH

This issue occurs because the full compiled stack size exceeds the FLASH size of the Raspberry Pi Pico W. This can be resolved by reducing the number of compiled 
features in the stack. Instructions for how to do this can be found in the CAS BACnet Stack Quick Start Guide in the docs folder of the stack.


CHANGELOG.md
# Change Log

## Version 0.0.x

### 0.0.1.x (2023-Oct-10)

- Initial version


lwipopts.h
#ifndef _LWIPOPTS_H
#define _LWIPOPTS_H

// From https://github.com/raspberrypi/pico-examples/blob/master/pico_w/wifi/lwipopts_examples_common.h
// Common settings used in most of the pico_w examples
// (see https://www.nongnu.org/lwip/2_1_x/group__lwip__opts.html for details)

// allow override in some examples
#ifndef NO_SYS
#define NO_SYS                      1
#endif
// allow override in some examples
#ifndef LWIP_SOCKET
#define LWIP_SOCKET                 0
#endif
#if PICO_CYW43_ARCH_POLL
#define MEM_LIBC_MALLOC             1
#else
// MEM_LIBC_MALLOC is incompatible with non polling versions
#define MEM_LIBC_MALLOC             0
#endif
#define MEM_ALIGNMENT               4
#define MEM_SIZE                    4000
#define MEMP_NUM_TCP_SEG            32
#define MEMP_NUM_ARP_QUEUE          10
#define PBUF_POOL_SIZE              24
#define LWIP_ARP                    1
#define LWIP_ETHERNET               1
#define LWIP_ICMP                   1
#define LWIP_RAW                    1
#define TCP_WND                     (8 * TCP_MSS)
#define TCP_MSS                     1460
#define TCP_SND_BUF                 (8 * TCP_MSS)
#define TCP_SND_QUEUELEN            ((4 * (TCP_SND_BUF) + (TCP_MSS - 1)) / (TCP_MSS))
#define LWIP_NETIF_STATUS_CALLBACK  1
#define LWIP_NETIF_LINK_CALLBACK    1
#define LWIP_NETIF_HOSTNAME         1
#define LWIP_NETCONN                0
#define MEM_STATS                   0
#define SYS_STATS                   0
#define MEMP_STATS                  0
#define LINK_STATS                  0
// #define ETH_PAD_SIZE                2
#define LWIP_CHKSUM_ALGORITHM       3
#define LWIP_DHCP                   1
#define LWIP_IPV4                   1
#define LWIP_TCP                    1
#define LWIP_UDP                    1
#define LWIP_DNS                    1
#define LWIP_TCP_KEEPALIVE          1
#define LWIP_NETIF_TX_SINGLE_PBUF   1
#define DHCP_DOES_ARP_CHECK         0
#define LWIP_DHCP_DOES_ACD_CHECK    0

#ifndef NDEBUG
#define LWIP_DEBUG                  1
#define LWIP_STATS                  1
#define LWIP_STATS_DISPLAY          1
#endif

#define ETHARP_DEBUG                LWIP_DBG_OFF
#define NETIF_DEBUG                 LWIP_DBG_OFF
#define PBUF_DEBUG                  LWIP_DBG_OFF
#define API_LIB_DEBUG               LWIP_DBG_OFF
#define API_MSG_DEBUG               LWIP_DBG_OFF
#define SOCKETS_DEBUG               LWIP_DBG_OFF
#define ICMP_DEBUG                  LWIP_DBG_OFF
#define INET_DEBUG                  LWIP_DBG_OFF
#define IP_DEBUG                    LWIP_DBG_OFF
#define IP_REASS_DEBUG              LWIP_DBG_OFF
#define RAW_DEBUG                   LWIP_DBG_OFF
#define MEM_DEBUG                   LWIP_DBG_OFF
#define MEMP_DEBUG                  LWIP_DBG_OFF
#define SYS_DEBUG                   LWIP_DBG_OFF
#define TCP_DEBUG                   LWIP_DBG_OFF
#define TCP_INPUT_DEBUG             LWIP_DBG_OFF
#define TCP_OUTPUT_DEBUG            LWIP_DBG_OFF
#define TCP_RTO_DEBUG               LWIP_DBG_OFF
#define TCP_CWND_DEBUG              LWIP_DBG_OFF
#define TCP_WND_DEBUG               LWIP_DBG_OFF
#define TCP_FR_DEBUG                LWIP_DBG_OFF
#define TCP_QLEN_DEBUG              LWIP_DBG_OFF
#define TCP_RST_DEBUG               LWIP_DBG_OFF
#define UDP_DEBUG                   LWIP_DBG_OFF
#define TCPIP_DEBUG                 LWIP_DBG_OFF
#define PPP_DEBUG                   LWIP_DBG_OFF
#define SLIP_DEBUG                  LWIP_DBG_OFF
#define DHCP_DEBUG                  LWIP_DBG_OFF

#endif

CASBACnetStackExampleConstants.h
/*
 * RP2040 BACnet Server Example 
 * ----------------------------------------------------------------------------
 * CASBACnetStackExampleConstants.h
 * 
 * This is a fully static class that contains all the constants used by the example.
 * Includes Object Types, Property Identifiers, etc.
 * 
 * Created by: Steven Smethurst
*/

#ifndef __CASBACnetStackExampleConstants_h__
#define __CASBACnetStackExampleConstants_h__

#include "datatypes.h"

class CASBACnetStackExampleConstants {

public:

	// CAS BACnet Stack network type
	static const uint8_t NETWORK_TYPE_IP = 0;
	static const uint8_t NETWORK_TYPE_MSTP = 1;


	// General Constants
	static const uint32_t NETWORK_PORT_LOWEST_PROTOCOL_LAYER = 4194303;

	// Object Types
	static const uint16_t OBJECT_TYPE_ANALOG_INPUT = 0;
	static const uint16_t OBJECT_TYPE_ANALOG_OUTPUT = 1;
	static const uint16_t OBJECT_TYPE_ANALOG_VALUE = 2;
	static const uint16_t OBJECT_TYPE_BINARY_INPUT = 3;
	static const uint16_t OBJECT_TYPE_BINARY_OUTPUT = 4;
	static const uint16_t OBJECT_TYPE_BINARY_VALUE = 5;
	//static const uint16_t OBJECT_TYPE_CALENDAR = 6;
	//static const uint16_t OBJECT_TYPE_COMMAND = 7;
	static const uint16_t OBJECT_TYPE_DEVICE = 8;
	//static const uint16_t OBJECT_TYPE_EVENT_ENROLLMENT = 9;
	//static const uint16_t OBJECT_TYPE_FILE = 10;
	//static const uint16_t OBJECT_TYPE_GROUP = 11;
	//static const uint16_t OBJECT_TYPE_LOOP = 12;
	static const uint16_t OBJECT_TYPE_MULTI_STATE_INPUT = 13;
	static const uint16_t OBJECT_TYPE_MULTI_STATE_OUTPUT = 14;
	//static const uint16_t OBJECT_TYPE_NOTIFICATION_CLASS = 15;
	//static const uint16_t OBJECT_TYPE_PROGRAM = 16;
	//static const uint16_t OBJECT_TYPE_SCHEDULE = 17;
	//static const uint16_t OBJECT_TYPE_AVERAGING = 18;
	static const uint16_t OBJECT_TYPE_MULTI_STATE_VALUE = 19;
	static const uint16_t OBJECT_TYPE_TREND_LOG = 20;
	//static const uint16_t OBJECT_TYPE_LIFE_SAFETY_POINT = 21;
	//static const uint16_t OBJECT_TYPE_LIFE_SAFETY_ZONE = 22;
	//static const uint16_t OBJECT_TYPE_ACCUMULATOR = 23;
	//static const uint16_t OBJECT_TYPE_PULSE_CONVERTER = 24;
	//static const uint16_t OBJECT_TYPE_EVENT_LOG = 25;
	//static const uint16_t OBJECT_TYPE_GLOBAL_GROUP = 26;
	static const uint16_t OBJECT_TYPE_TREND_LOG_MULTIPLE = 27;
	//static const uint16_t OBJECT_TYPE_LOAD_CONTROL = 28;
	//static const uint16_t OBJECT_TYPE_STRUCTURED_VIEW = 29;
	//static const uint16_t OBJECT_TYPE_ACCESS_DOOR = 30;
	//static const uint16_t OBJECT_TYPE_TIMER = 31;
	//static const uint16_t OBJECT_TYPE_ACCESS_CREDENTIAL = 32;
	//static const uint16_t OBJECT_TYPE_ACCESS_POINT = 33;
	//static const uint16_t OBJECT_TYPE_ACCESS_RIGHTS = 34;
	//static const uint16_t OBJECT_TYPE_ACCESS_USER = 35;
	//static const uint16_t OBJECT_TYPE_ACCESS_ZONE = 36;
	//static const uint16_t OBJECT_TYPE_CREDENTIAL_DATA_INPUT = 37;
	//static const uint16_t OBJECT_TYPE_NETWORK_SECURITY = 38;
	static const uint16_t OBJECT_TYPE_BITSTRING_VALUE = 39;
	static const uint16_t OBJECT_TYPE_CHARACTERSTRING_VALUE = 40;
	//static const uint16_t OBJECT_TYPE_DATEPATTERN_VALUE = 41;
	static const uint16_t OBJECT_TYPE_DATE_VALUE = 42;
	//static const uint16_t OBJECT_TYPE_DATETIMEPATTERN_VALUE = 43;
	static const uint16_t OBJECT_TYPE_DATETIME_VALUE = 44;
	static const uint16_t OBJECT_TYPE_INTEGER_VALUE = 45;
	static const uint16_t OBJECT_TYPE_LARGE_ANALOG_VALUE = 46;
	static const uint16_t OBJECT_TYPE_OCTETSTRING_VALUE = 47;
	static const uint16_t OBJECT_TYPE_POSITIVE_INTEGER_VALUE = 48;
	//static const uint16_t OBJECT_TYPE_TIMEPATTERN_VALUE = 49;
	static const uint16_t OBJECT_TYPE_TIME_VALUE = 50;
	//static const uint16_t OBJECT_TYPE_NOTIFICATION_FORWARDER = 51;
	//static const uint16_t OBJECT_TYPE_ALERT_ENROLLMENT = 52;
	//static const uint16_t OBJECT_TYPE_CHANNEL = 53;
	//static const uint16_t OBJECT_TYPE_LIGHTING_OUTPUT = 54;
	//static const uint16_t OBJECT_TYPE_BINARY_LIGHTING_OUTPUT = 55;
	static const uint16_t OBJECT_TYPE_NETWORK_PORT = 56;
	//static const uint16_t OBJECT_TYPE_ELEVATOR_GROUP = 57;
	//static const uint16_t OBJECT_TYPE_ESCALATOR = 58;
	//static const uint16_t OBJECT_TYPE_LIFT = 59;

	// Property Identifiers
	static const uint32_t PROPERTY_IDENTIFIER_ALL = 8;
	static const uint32_t PROPERTY_IDENTIFIER_COV_INCURMENT = 22;
	static const uint32_t PROPERTY_IDENTIFIER_DAY_LIGHT_SAVINGS_STATUS = 24;
	static const uint32_t PROPERTY_IDENTIFIER_DESCRIPTION = 28;
	static const uint32_t PROPERTY_IDENTIFIER_LOCAL_DATE = 56;
	static const uint32_t PROPERTY_IDENTIFIER_LOCAL_TIME = 57;
	static const uint32_t PROPERTY_IDENTIFIER_NUMBER_OF_STATES = 74;
	static const uint32_t PROPERTY_IDENTIFIER_OBJECT_NAME = 77;	
	static const uint32_t PROPERTY_IDENTIFIER_PRESENT_VALUE = 85;
	static const uint32_t PROPERTY_IDENTIFIER_PRIORITY_ARRAY = 87;
	static const uint32_t PROPERTY_IDENTIFIER_RELIABILITY = 103;
	static const uint32_t PROPERTY_IDENTIFIER_STATE_TEXT = 110;
	static const uint32_t PROPERTY_IDENTIFIER_STATUS_FLAGS = 111;
	static const uint32_t PROPERTY_IDENTIFIER_SYSTEM_STATUS = 112;
	static const uint32_t PROPERTY_IDENTIFIER_UNITS = 117;
	static const uint32_t PROPERTY_IDENTIFIER_UTC_OFFSET = 119;
	static const uint32_t PROPERTY_IDENTIFIER_BIT_TEXT = 343;
	static const uint32_t PROPERTY_IDENTIFIER_APPLICATION_SOFTWARE_VERSION = 12;
	static const uint32_t PROPERTY_IDENTIFIER_RELINQUISH_DEFAULT = 104;

	
	

	static const uint32_t PROPERTY_IDENTIFIER_MAX_PRES_VALUE = 65;
	static const uint32_t PROPERTY_IDENTIFIER_MIN_PRES_VALUE = 69;

	// Network Port Property Identifiers
	static const uint32_t PROPERTY_IDENTIFIER_IP_ADDRESS = 400;
	static const uint32_t PROPERTY_IDENTIFIER_IP_DEFAULT_GATEWAY = 401;
	static const uint32_t PROPERTY_IDENTIFIER_IP_DNS_SERVER = 406;
	static const uint32_t PROPERTY_IDENTIFIER_IP_SUBNET_MASK = 411;
	static const uint32_t PROPERTY_IDENTIFIER_BACNET_IP_UDP_PORT = 412;
	static const uint32_t PROPERTY_IDENTIFIER_BBMD_ACCEPT_FD_REGISTRATIONS = 413;
	static const uint32_t PROPERTY_IDENTIFIER_BBMD_BROADCAST_DISTRIBUTION_TABLE = 414;
	static const uint32_t PROPERTY_IDENTIFIER_BBMD_FOREIGN_DEVICE_TABLE = 415;
	static const uint32_t PROPERTY_IDENTIFIER_CHANGES_PENDING = 416;
	static const uint32_t PROPERTY_IDENTIFIER_FD_BBMD_ADDRESS = 418;
	static const uint32_t PROPERTY_IDENTIFIER_FD_SUBSCRIPTION_LIFETIME = 419;
	static const uint32_t PROPERTY_IDENTIFIER_LINK_SPEED = 420;
	static const uint32_t PROPERTY_IDENTIFIER_MAC_ADDRESS = 423;
	
	// Services Supported
	static const uint8_t SERVICE_SUBSCRIBE_COV = 5;
	static const uint8_t SERVICE_CREATE_OBJECT = 10;
	static const uint8_t SERVICE_DELETE_OBJECT = 11;
	static const uint8_t SERVICE_READ_PROPERTY_MULTIPLE = 14;
	static const uint8_t SERVICE_WRITE_PROPERTY = 15;
	static const uint8_t SERVICE_WRITE_PROPERTY_MULTIPLE = 16;
	static const uint8_t SERVICE_DEVICE_COMMUNICATION_CONTROL = 17;
	static const uint8_t SERVICE_CONFIRMED_TEXT_MESSAGE = 19;
	static const uint8_t SERVICE_REINITIALIZE_DEVICE = 20;
	static const uint8_t SERVICE_I_AM = 26;
	static const uint8_t SERVICE_I_HAVE = 27;
	static const uint8_t SERVICE_UNCONFIRMED_TEXT_MESSAGE = 31;
	static const uint8_t SERVICE_TIME_SYNCHRONIZATION = 32;
	static const uint8_t SERVICE_READ_RANGE = 35;
	static const uint8_t SERVICE_UTC_TIME_SYNCHRONIZATION = 36;
	static const uint8_t SERVICE_SUBSCRIBE_COV_PROPERTY = 38;
	

	// Error Codes
	static const uint8_t ERROR_MISSING_REQUIRED_PARAMETER = 16;
	static const uint8_t ERROR_NO_SPACE_TO_WRITE_PROPERTY = 20;
	static const uint8_t ERROR_PASSWORD_FAILURE = 26;
	static const uint8_t ERROR_VALUE_OUT_OF_RANGE = 37;
	static const uint8_t ERROR_OPTIONAL_FUNCTIONALITY_NOT_SUPPORTED = 45;
	static const uint8_t ERROR_INVALID_CONFIGURATION_DATA = 46;

	// Network Port FdBBmdAddressHostType
	static const uint8_t HOST_TYPE_NONE = 0;
	static const uint8_t HOST_TYPE_IPADDRESS = 1;
	static const uint8_t HOST_TYPE_NAME = 2;

	// Network Port FdBbmdAddressOffset
	static const uint8_t FD_BBMD_ADDRESS_HOST = 1;
	static const uint8_t FD_BBMD_ADDRESS_PORT = 2;

	// Network Type
	static const uint8_t NETWORK_TYPE_BACNET_IP = 0;
	static const uint8_t NETWORK_TYPE_IPV4 = 5;

	// Protocol Level
	static const uint8_t PROTOCOL_LEVEL_BACNET_APPLICATION = 2;

	// Priority Array
	static const uint8_t MAX_BACNET_PRIORITY = 16;

	// Trend Log Buffer Size
	static const uint32_t MAX_TREND_LOG_MAX_BUFFER_SIZE = 100;

	// Data types 
	static const uint32_t DATA_TYPE_NULL = 0;
	static const uint32_t DATA_TYPE_BOOLEAN = 1;
	static const uint32_t DATA_TYPE_UNSIGNED_INTEGER = 2;
	static const uint32_t DATA_TYPE_SIGNED_INTEGER = 3;
	static const uint32_t DATA_TYPE_REAL = 4;
	static const uint32_t DATA_TYPE_DOUBLE = 5;
	static const uint32_t DATA_TYPE_OCTET_STRING = 6;
	static const uint32_t DATA_TYPE_CHARACTER_STRING = 7;
	static const uint32_t DATA_TYPE_BIT_STRING = 8;
	static const uint32_t DATA_TYPE_ENUMERATED = 9;
	static const uint32_t DATA_TYPE_DATE = 10;
	static const uint32_t DATA_TYPE_TIME = 11;
	static const uint32_t DATA_TYPE_BACNET_OBJECT_IDENTIFIER = 12;
	static const uint32_t DATA_TYPE_DATETIME = 27;

	// Temperature enum
	static const uint32_t TEMPERATURE_ENUM_DEGREES_CELSIUS = 62;
	static const uint32_t TEMPERATURE_ENUM_DEGREES_FAHRENHEIT = 64;

	// Reinitialized State
	static const uint8_t REINITIALIZED_STATE_WARM_START = 1;
	static const uint8_t REINITIALIZED_STATE_ACTIVATE_CHANGES = 7;

	// Debug Message Type
	static const uint8_t BACNET_DEBUG_LOG_TYPE_ERROR = 0;
	static const uint8_t BACNET_DEBUG_LOG_TYPE_INFO = 1;
};

#endif // __CASBACnetStackExampleConstants_h__